src/auth/ui/vue/show.ts

```diff
@@ -332,18 +358,232 @@ export class AuthWebview extends VueWebview {
+     *
+     * Eg: Builder ID for CodeWhisperer, Credentials for AWS Explorer
+     */
+    private previousAuthType: CredentialSourceId | undefined
+    private previousFeatureType: AuthUiElement | undefined
+
+    /**
+     * This function is called whenever some sort of interaction with an auth form happens in
+     * the webview. This helps keeps track of the auth form that was last interacted with.
+     *
+     * If a user starts interacting with a subsequent form this will emit a metric
+     * related to an **unsuccessful** connection attempt to the previous auth form.
+     */
+    startAuthFormInteraction(featureType: AuthUiElement, authType: CredentialSourceId) {
+        // Check if a previous auth interaction existed and that the new one is not the same as it
+        if (
+            this.previousFeatureType !== undefined &&
+            this.previousAuthType !== undefined &&
+            (this.previousFeatureType !== featureType || this.previousAuthType !== authType)
+        ) {
+            // At this point a user WAS previously interacting with a different auth form
+            // and started interacting with a new one (hence the new feature + auth type).
+            // We can now indicate that the previous one was cancelled and clear out any state values
+            this.endExistingAuthFormInteraction(this.previousFeatureType, this.previousAuthType)
+        }
+
+        this.previousAuthType = authType
+        this.previousFeatureType = featureType
+    }
+
+    /**
+     * The metric for when an auth form that was unsuccessfully interacted with is
+     * done being interacted with
+     */
+    private endExistingAuthFormInteraction(featureType: AuthUiElement, authType: CredentialSourceId) {
+        this.emitAuthAttempt({
+            authType,
+            featureType,
+            result: 'Cancelled',
+            invalidFields: this.invalidFields,
+            reason: 'incompleteAuthForm',
+        })
+
+        this.invalidFields = undefined
+    }
+
+    /**
+     * This is run when an auth form interaction was successful.
+     *
+     * We do this so subsequent auth form interaction checks don't
+     * assume that this was cancelled, since we look back at the following
+     * properties to see if an auth form was not successfully completed.
+     */
+    private successfulAuthFormInteraction() {
+        this.previousAuthType = undefined
+        this.previousFeatureType = undefined
+    }
+
+    /**
+     * The metric when the webview is opened.
+     *
+     * Think of this as a snapshot of the state at the start.
+     */
+    async emitOpened() {
+        telemetry.auth_addConnection.emit({
+            source: this.getSource() ?? '',
+            reason: 'opened',
+            authConnectionsCount: await this.getConnectionCount(),
+            authEnabledAreas: builderCommaDelimitedString(this.getUnlockedFeatures()),
+            result: 'Cancelled',
+        })
+    }
```

Comment:
Not a fan of using `Cancelled` when nothing was cancelled. It would make more sense to attach this data to one of the entrypoint events.

Reply:
Right now the `result` field is required, but I don't think we need it to be. `Cancelled` is a placeholder for now, since if we see `reason: 'opened'` we know that `result` is irrelevant.